include/eve/detail/shuffle_v2/simd/riscv/shuffle_l2.hpp

```diff
@@ -0,0 +1,53 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+namespace eve::detail
+{
+template<typename P, scalar_value T, typename N, std::ptrdiff_t G>
+EVE_FORCEINLINE auto
+shuffle_l2_(EVE_SUPPORTS(rvv_), P p, fixed<G> g, wide<T, N> x)
+{
+  using int_ty = as_integer_t<T, unsigned>;
+  if constexpr( G == N::value )
+  {
+    wide<as_integer_t<T, unsigned>, N> offsets(P::idxs);
+    wide<T, N>                         res = __riscv_vrgather_tu(x, x, offsets.storage(), N::value);
+    return res;
+  }
+  else
+  {
+    std::array<int_ty, N::value> ids;
+    for( auto id = 0; id < N::value; ++id )
+    {
+      size_t id_from_pattern = id / G;
+      ids[id]                = P::idxs[id_from_pattern] * G + id % G;
+    }
+    wide<as_integer_t<T, unsigned>, N> offsets(ids);
+    wide<T, N>                         res = __riscv_vrgather_tu(x, x, offsets.storage(), N::value);
```

Comment:
Gather is level 3 - it needs a mask. 

Here is how you do it: https://github.com/jfalcou/eve/blob/5c21a5b76ccbf5c00cccaa852421c0746df3990d/include/eve/detail/shuffle_v2/simd/arm/sve/shuffle_l3.hpp#L30

Let me know if that doesn't make sense, and I'll help more.